ocis/pkg/command/migrate.go

```diff
@@ -0,0 +1,283 @@
+package command
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+
+	collaboration "github.com/cs3org/go-cs3apis/cs3/sharing/collaboration/v1beta1"
+	"github.com/cs3org/reva/v2/pkg/publicshare"
+	publicregistry "github.com/cs3org/reva/v2/pkg/publicshare/manager/registry"
+	"github.com/cs3org/reva/v2/pkg/share"
+	"github.com/cs3org/reva/v2/pkg/share/manager/registry"
+	sharing "github.com/owncloud/ocis/v2/extensions/sharing/pkg/config"
+	"github.com/owncloud/ocis/v2/ocis-pkg/config"
+	"github.com/owncloud/ocis/v2/ocis-pkg/config/parser"
+	"github.com/owncloud/ocis/v2/ocis/pkg/register"
+	"github.com/rs/zerolog"
+	"github.com/urfave/cli/v2"
+)
+
+// Migrate is the entrypoint for the Migrate command.
+func Migrate(cfg *config.Config) *cli.Command {
+	return &cli.Command{
+		Name:     "migrate",
+		Usage:    "migrate data from an existing instance to a new version",
+		Category: "migration",
+		Before: func(c *cli.Context) error {
+			if err := parser.ParseConfig(cfg, true); err != nil {
+				fmt.Printf("%v", err)
+				return err
+			}
+			return nil
+		},
+		Subcommands: []*cli.Command{
+			MigrateShares(cfg),
+			MigratePublicShares(cfg),
+		},
+	}
+}
+
+func init() {
+	register.AddCommand(Migrate)
+}
+
+func MigrateShares(cfg *config.Config) *cli.Command {
+	return &cli.Command{
+		Name:  "shares",
+		Usage: "migrates shares from the previous to the new manager",
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:  "from",
+				Value: "json",
+				Usage: "Share manager to export the data from",
+			},
+			&cli.StringFlag{
+				Name:  "to",
+				Value: "cs3",
+				Usage: "Share manager to import the data to",
+			},
+		},
+		Before: func(c *cli.Context) error {
+			err := parser.ParseConfig(cfg, true)
+			if err != nil {
+				fmt.Printf("%v", err)
+				os.Exit(1)
+			}
+			return err
+		},
+		Action: func(c *cli.Context) error {
+			log := zerolog.New(zerolog.ConsoleWriter{Out: os.Stderr}).With().Timestamp().Logger()
+			ctx := log.WithContext(context.Background())
+			rcfg := revaShareConfig(cfg.Sharing)
+			oldDriver := c.String("from")
+			newDriver := c.String("to")
+			shareChan := make(chan *collaboration.Share)
+			receivedShareChan := make(chan share.ReceivedShareWithUser)
+
+			f, ok := registry.NewFuncs[oldDriver]
+			if !ok {
+				log.Error().Msg("Unknown share manager type '" + oldDriver + "'")
+				os.Exit(1)
+			}
+			oldMgr, err := f(rcfg[oldDriver].(map[string]interface{}))
+			if err != nil {
+				log.Error().Err(err).Msg("failed to initiate source share manager")
+				os.Exit(1)
+			}
+			if _, ok := oldMgr.(share.DumpableManager); !ok {
```

Comment:
Any reason why we don't store the asserted `DumpableManager` into a var? Then we would not need to assert it again later...
(same question for the `LoadableManager`)